src/Tactic/Cong.agda

```diff
@@ -0,0 +1,185 @@
+
+      ex-args hole-i [] = return []
+      ex-args hole-i (arg v t ∷ as) = do
+        t ← ex hole-i t
+        as ← ex-args hole-i as
+        return $ arg v t ∷ as
+
+      ex-clauses hole-i [] = return []
+      ex-clauses hole-i (clause ps t ∷ cs) = do
+        t ← ex (hole-i + pattern-args-size ps) t
+        cs ← ex-clauses hole-i cs
+        return $ clause ps t ∷ cs
+      ex-clauses hole-i (absurd-clause ps ∷ cs) = do
+        cs ← ex-clauses hole-i cs
+        return $ absurd-clause ps ∷ cs
+
+open import Reflection hiding (map-Args ; _≟_)
+
+data RelSide : Set 0ℓ where
+  lhs : RelSide
+  rhs : RelSide
+
+private
+  Sort→Level : Sort → TC Level
+  Sort→Level unknown = typeError (strErr "Tactic.Cong: Could not determine sort of return type" ∷ [])
+  Sort→Level (set ℓ) = unquoteTC ℓ
+  Sort→Level (lit n) = return (for-n n)
+    where
+      for-n : ℕ → Level
+      for-n ℕ.zero = 0ℓ
+      for-n (suc n) = lsuc (for-n n)
+
+  -- for idiom brackets
+  pure : Term → TC Term
+  pure t = return t
+
+  -- for idiom brackets
+  _<*>_ : TC Term → Arg Term → TC Term
+  f <*> x = do
+    f ← f
+    case f ⟨ apply ⟩ x of λ {
+        (err e) → typeError $ strErr "Tactic.Cong:" ∷ e ;
+        (ok t) → return t
+      }
+
+  ⌊⌋' : ∀ {ℓ} {A : Set ℓ} {x y : A} → RelSide → x ≡ y → Term → TC ⊤
+  ⌊⌋' {ℓ = aℓ} {A = A} {x = x} {y = y} rel-side x≡y hole = do
+    hole-T ← inferType hole
+    agda-sort hole-sort ← withNormalisation true $ inferType hole-T
+      where other → typeError $ strErr "Tactic.Cong: expected a sort, got" ∷ termErr other ∷ []
+    bℓ ← Sort→Level hole-sort
+    just (fx ∷ fy ∷ _) ← return $ term-vis-args hole-T
+      where _ → typeError $ strErr "Tactic.Cong: unexpected form for relation in" ∷ termErr hole-T ∷ []
+    cong ← quoteTC {A = congT aℓ bℓ} cong
+    Set-bℓ ← quoteTC (Set bℓ)
+    A→ ← quoteTC (λ (B : Set bℓ) → (A → B))
+    A ← quoteTC A
+    B ← checkType unknown Set-bℓ
+    A→B ← ⦇ A→ (vArg B) ⦈
+    let chosen-side = case rel-side of λ { lhs → fx ; rhs → fy }
+    f , true ← return $ extract-f $ chosen-side
+      where _ , false → typeError $ strErr "Tactic.Cong: could not find ⌊_⌋ in" ∷ termErr chosen-side ∷ []
+    f ← checkType f A→B
+    x ← quoteTC x
+    y ← quoteTC y
+    x≡y ← quoteTC x≡y
+    out ← ⦇ cong (hArg A) (hArg B) (vArg f) (hArg x) (hArg y) (vArg x≡y) ⦈
+    unify hole out
+    where
+      congT : ∀ aℓ bℓ → Set (lsuc (lmax aℓ bℓ))
+      congT aℓ bℓ =
+        {A : Set aℓ}
+        {B : Set bℓ}
+        (f : A → B)
+        {x y : A}
+        (p : x ≡ y)
+        → f x ≡ f y
+
+macro
+  {- call cong, inferring the function argument from the use of ⌊_⌋. The side of
+     the relation inspected is determined by the RelSide parameter
+  -}
+  ⌊⌋ : ∀ {ℓ} {A : Set ℓ} {x y : A} → RelSide → x ≡ y → Term → TC ⊤
+  ⌊⌋ side = ⌊⌋' side
+
+  {- Convenience macro equivalent to using ⌊⌋ within _≡⟨_⟩_ -}
+  _≡⌊_⌋_ : ∀ {ℓ} {A : Set ℓ} → Term → {x y : A} → x ≡ y → Term → Term → TC ⊤
```

Comment:
It's incredibly neat that we don't have to add it to every version of equational reasoning :tada:

Reply:
We should probably also have the `sym`'d one.

Reply:
Has syntax been agreed for that yet?

Reply:
Nope, so maybe leave it out for the moment, and we'll add it in later when we've decided.